torchtext/data/datasets_utils.py

```diff
@@ -279,18 +280,39 @@ def _download_extract_validate(root, url, url_md5, downloaded_file, extracted_fi
     return extracted_file
 
 
+class _RawTextMapDataset(torch.utils.data.Dataset):
+    """Helper class to convert IterableDatatset to Dataset"""
+
+    def __init__(self, raw_text_iter_data):
+        """Initialized with _RawTextIterableDataset"""
+        self.description = raw_text_iter_data.description
+        self._data = [d for d in raw_text_iter_data]
+
+    def __len__(self):
+        return len(self._data)
+
+    def __getitem__(self, idx):
+        return self._data[idx]
+
+
 class _RawTextIterableDataset(torch.utils.data.IterableDataset):
-    """Defines an abstraction for raw text iterable datasets.
-    """
+    """Defines an abstraction for raw text iterable datasets."""
 
-    def __init__(self, description, full_num_lines, iterator):
-        """Initiate the dataset abstraction.
-        """
+    def __init__(self, description, full_num_lines, iterator, offset=0):
+        """Initiate the dataset abstraction."""
         super(_RawTextIterableDataset, self).__init__()
+        assert (
+            offset < full_num_lines
+        ), "offset {} should be less than the number of lines {} in dataset".format(
+            offset, full_num_lines
+        )
+        assert offset >= 0, "offset must be non-negative but {} is given".format(offset)
         self.description = description
         self.full_num_lines = full_num_lines
         self._iterator = iterator
-        self.num_lines = full_num_lines
+        self.num_lines = full_num_lines - offset
+        for _ in range(offset):
```

Comment:
Hi @parmeet, thanks for taking this issue. I tried this method before when training XLM-R with CC100. In practice, it does take time to skip lines and for a very large dataset, it's not efficient. What I really want is the real offset, something similar to the legacy EnWik9 dataset. Here are some [code pointers](https://github.com/pytorch/text/blob/ba0bf52fd0f7981407ec7479918a5d34b00d2211/torchtext/legacy/datasets/unsupervised_learning.py#L112-L115). Let me know if this makes sense to you.

Reply:
> Hi @parmeet, thanks for taking this issue. I tried this method before when training XLM-R with CC100. In practice, it does take time to skip lines and for a very large dataset, it's not efficient. What I really want is the real offset, something similar to the legacy EnWik9 dataset. Here are some [code pointers](https://github.com/pytorch/text/blob/ba0bf52fd0f7981407ec7479918a5d34b00d2211/torchtext/legacy/datasets/unsupervised_learning.py#L112-L115). Let me know if this makes sense to you.
> 
> Technically, we can generate those offsets offline and save them so users can access them efficiently. wdyt? cc @cpuhrsch

Hi @zhangguanheng66 , Thanks for the proposal. I think it's a good idea. Although I am not sure, what should be viable solution for offsets storage. I could think of following:
1) Provide them as package binary.
Pros: Instant availability
Cons: Package size
2) Store it in public cloud storage and download it on-demand.
Pros: Almost Instant availability
Cons: maintenance/dependence on public cloud.
3) Generate it once when the user requests the data and then store it in local cache for later use.
Pros: We could potentially avoid cons of 1 and 2
cons: Need to iterate over dataset at-least once before we return the iterator to user.

What are your thoughts on this?

Reply:
I think the options here will depends on the size of the datasets. We can do an experiment with WikiText103/AmazonFullReview and see how this will affect the performance.

Reply:
Typically a user would split data at particular offset and most likely stick with it in consecutive uses/epochs. Hence it may not be necessary to provide offsets for all the lines. But we do not know in advance what offset the user would want. Hence we can decide to create this offset on-the-fly when user request the iterator at given offset. 

IMHO, I won't recommend option 1 and 2 above for reasons stated in the cons and for unnecessary storage at all the offsets. Option 3 seems like a good trade-off with efficiency and storage requirements. For option 3 the workflow could look like this:
1) If user do not provide offset, we just return the iterator as we do now (zero-cost)
2) If user do request data with offset, we first check if offset is cached or not. 
    2.1) If yes, we simply read it from cache and return the offset iterator (zero-cost).
    2.2) If not, then we first iterate over dataset (one-time cost) and store the corresponding offset in cache (minimal storage/maintenance requirements) and then go to step 2.1

This way for any consecutive call to data iterator, we do not incur any additional run-time cost and at the same time need not to worry about potential storage for offsets.

Reply:
Another option is to store a few offset values (say every 10,000 lines) and their corresponding byte offset. Then we can use the very efficient [seek](https://man7.org/linux/man-pages/man2/lseek.2.html) in cases where the underlying text data is a file. So if a user requests offset 12,345 we'll first seek to line 10,000 based on the stored byte value and then forward the iterator by 2,345. Of course we can vary the values here (store a larger array of offset metadata, etc.).

Reply:
Thanks @cpuhrsch for the proposal. This sounds like a pretty decent approach to me!